File: frontend/src/engine.ts
import type { EvaluationInput, EvaluationResult, Methodology, ProcessModel } from './types'

function asNumber(value: unknown): number {
  const n = Number(value)
  return Number.isNaN(n) ? 0 : n
}

function clamp(value: number, min = 0, max = 100): number {
  let n = asNumber(value)
  if (n < min) n = min
  if (n > max) n = max
  return n
}

function linearClamp(
  value: number,
  inMin: number,
  inMax: number,
  outMin = 0,
  outMax = 100,
): number {
  if (inMin >= inMax) return outMin
  const normalized = (clamp(value, inMin, inMax) - inMin) / (inMax - inMin)
  return outMin + normalized * (outMax - outMin)
}

function roundScore(value: number, decimals = 1): number {
  return Math.round(value * 10 ** decimals) / 10 ** decimals
}

function classify(methodology: Methodology, score: number): { label: string; description: string } {
  const bands = [...methodology.classification.bands].sort((a, b) => b.min - a.min)
  for (const band of bands) {
    if (score >= band.min) return { label: band.label, description: band.description }
  }
  return { label: 'Sin clasificación', description: '' }
}

function computeConfidence(
  methodology: Methodology,
  model: ProcessModel,
  plan: string,
  evidenceQuality: number,
): number {
  const equipmentCap = model.equipment_capability
  const planCap = methodology.confidence.plan_caps[plan.toLowerCase()] || 0
  const evidenceCap = clamp(evidenceQuality, 0, methodology.confidence.range.max)
  return Math.min(equipmentCap, planCap, evidenceCap)
}

function computeIntegrityScore(
  methodology: Methodology,
  values: Record<string, number | string>,
): number {
  let total = 0
  for (const [key, component] of Object.entries(methodology.integrity.components)) {
    const value = asNumber(values[key] ?? 0)
    total += clamp(value) * component.weight
  }
  return total
}

function computeProcessScore(model: ProcessModel, values: Record<string, number | string>): number {
  let total = 0
  for (const field of model.fields) {
    const value = asNumber(values[field.id] ?? 0)
    total += clamp(value) * field.weight
  }
  return total
}

function computeTotalPenalties(penalties: { value: number | string }[]): number {
  return clamp(
    penalties.reduce((sum, p) => sum + asNumber(p.value), 0),
    0,
    100,
  )
}

export function evaluate(input: EvaluationInput, methodology: Methodology): EvaluationResult {
  const quality = methodology.quality
  const qualityValue = clamp(asNumber(input.quality_input), quality.input_range.min, quality.input_range.max)
  const qualityScore = linearClamp(
    qualityValue,
    quality.normalization.min_input,
    quality.normalization.max_input,
    quality.normalization.min_output,
    quality.normalization.max_output,
  )

  const processModel = methodology.process.models[input.equipment_model]
  if (!processModel) {
    throw new Error(`Modelo de proceso '${input.equipment_model}' no encontrado`)
  }

  const processScore = computeProcessScore(processModel, input.process_values)
  const integrityScore = computeIntegrityScore(methodology, input.integrity_values)
  const totalPenalties = computeTotalPenalties(input.penalties)

  const rawScore =
    qualityScore * methodology.weights.quality +
    processScore * methodology.weights.process +
    integrityScore * methodology.weights.integrity -
    totalPenalties

  const totalScore = clamp(rawScore, methodology.score_range.min, methodology.score_range.max)
  const classification = classify(methodology, totalScore)
  const confidenceLevel = computeConfidence(
    methodology,
    processModel,
    input.origin_plan,
    asNumber(input.evidence_quality),
  )

  const now = new Date().toISOString()
  const id = Math.floor(Date.now() / 1000)

  const processInputs: Record<string, number> = {}
  for (const field of processModel.fields) {
    processInputs[field.id] = roundScore(clamp(asNumber(input.process_values[field.id] ?? 0)))
  }

  const integrityComponents: Record<string, number> = {}
  for (const key of Object.keys(methodology.integrity.components)) {
    integrityComponents[key] = roundScore(clamp(asNumber(input.integrity_values[key] ?? 0)))
  }

  return {
    id,
    public_id: `NS-${id}-${Math.floor(Math.random() * 1000)}`,
    product: input.product,
    methodology_id: methodology.id,
    methodology_version: methodology.version,
    status: 'calculated',
    lot_id: input.lot_id || undefined,
    producer: input.producer || undefined,
    farm: input.farm || undefined,
    country: input.country || undefined,
    region: input.region || undefined,
    geo_latitude: input.geo_latitude ? asNumber(input.geo_latitude) : undefined,
    geo_longitude: input.geo_longitude ? asNumber(input.geo_longitude) : undefined,
    variety: input.variety || undefined,
    harvest_date: input.harvest_date || undefined,
    process_start_date: input.process_start_date || undefined,
    process_end_date: input.process_end_date || undefined,
    equipment_model: input.equipment_model,
    origin_plan: input.origin_plan,
    evidence_quality: asNumber(input.evidence_quality),
    protocol: input.protocol || undefined,
    quality_input: asNumber(input.quality_input),
    process_values: Object.fromEntries(
      Object.entries(input.process_values).map(([k, v]) => [k, roundScore(asNumber(v))]),
    ),
    integrity_values: Object.fromEntries(
      Object.entries(input.integrity_values).map(([k, v]) => [k, roundScore(asNumber(v))]),
    ),
    penalties: input.penalties.map((p) => ({ ...p, value: roundScore(asNumber(p.value)) })),
    quality_score: roundScore(qualityScore),
    process_score: roundScore(processScore),
    integrity_score: roundScore(integrityScore),
    total_penalties: roundScore(totalPenalties),
    nebula_score: roundScore(totalScore),
    confidence_level: confidenceLevel,
    classification: classification.label,
    interpretation: classification.description,
    components: {
      quality: {
        label: methodology.quality.label,
        raw_input: roundScore(qualityValue, 2),
        normalized_score: roundScore(qualityScore),
        weight: methodology.weights.quality,
        weighted_score: roundScore(qualityScore * methodology.weights.quality),
        warning: methodology.quality.warning,
      },
      process: {
        label: methodology.process.label,
        model: processModel.label,
        model_formula: processModel.formula,
        score: roundScore(processScore),
        weight: methodology.weights.process,
        weighted_score: roundScore(processScore * methodology.weights.process),
        inputs: processInputs,
      },
      integrity: {
        label: methodology.integrity.label,
        score: roundScore(integrityScore),
        weight: methodology.weights.integrity,
        weighted_score: roundScore(integrityScore * methodology.weights.integrity),
        components: integrityComponents,
      },
      penalties: {
        total: roundScore(totalPenalties),
        items: input.penalties,
      },
    },
    created_at: now,
    updated_at: now,
  }
}

File: backend/app/core/engine.py
from dataclasses import dataclass
from typing import Any

from backend.app.core.confidence import compute_confidence
from backend.app.core.exceptions import ValidationError
from backend.app.core.integrity import compute_integrity_score
from backend.app.core.methodology import Methodology, ProcessModelSpec
from backend.app.core.normalization import as_number, clamp, linear_clamp, round_score
from backend.app.core.penalty import Penalty, compute_total_penalties


@dataclass
class ScoreResult:
    product: str
    methodology_id: str
    methodology_version: str
    quality_score: float
    process_score: float
    integrity_score: float
    penalties: float
    total_score: float
    confidence_level: int
    classification: str
    interpretation: str
    components: dict[str, Any]


def _validate_and_normalize_quality(methodology: Methodology, raw_value: float) -> float:
    quality = methodology.quality
    value = clamp(as_number(raw_value), quality.input_range.min, quality.input_range.max)
    normalized = linear_clamp(
        value,
        quality.normalization.min_input,
        quality.normalization.max_input,
        quality.normalization.min_output,
        quality.normalization.max_output,
    )
    return normalized


def _compute_process_score(model: ProcessModelSpec, values: dict[str, float]) -> float:
    total = 0.0
    for field in model.fields:
        raw = values.get(field.id, 0.0)
        total += clamp(raw) * field.weight
    return total


def _classify(methodology: Methodology, score: float) -> tuple[str, str]:
    for band in sorted(methodology.classification.bands, key=lambda b: b.min, reverse=True):
        if score >= band.min:
            return band.label, band.description
    return "Sin clasificación", ""


def evaluate(
    methodology: Methodology,
    *,
    quality_input: float,
    process_values: dict[str, float],
    integrity_values: dict[str, float],
    penalties: list[Penalty],
    equipment_model: str,
    origin_plan: str,
    evidence_quality: int,
) -> ScoreResult:
    """Compute a Nebula Score® evaluation for a methodology and set of inputs."""
    quality_score = _validate_and_normalize_quality(methodology, quality_input)

    process_model = methodology.process.models.get(equipment_model)
    if process_model is None:
        available = ", ".join(methodology.process.models.keys())
        raise ValidationError(f"Process model '{equipment_model}' not found. Available: {available}")
    process_score = _compute_process_score(process_model, process_values)

    integrity_score = compute_integrity_score(methodology.integrity, integrity_values)

    total_penalties = compute_total_penalties(penalties, max_total=methodology.penalties.max_total)

    raw_score = (
        quality_score * methodology.weights.quality
        + process_score * methodology.weights.process
        + integrity_score * methodology.weights.integrity
        - total_penalties
    )
    total_score = clamp(raw_score, methodology.score_range.min, methodology.score_range.max)

    confidence_level = compute_confidence(
        methodology,
        equipment_model=equipment_model,
        plan=origin_plan,
        evidence_quality=evidence_quality,
    )

    classification, interpretation = _classify(methodology, total_score)

    components = {
        "quality": {
            "label": methodology.quality.label,
            "raw_input": round_score(as_number(quality_input), 2),
            "normalized_score": round_score(quality_score),
            "weight": methodology.weights.quality,
            "weighted_score": round_score(quality_score * methodology.weights.quality),
            "warning": methodology.quality.warning,
        },
        "process": {
            "label": methodology.process.label,
            "model": process_model.label,
            "model_formula": process_model.formula,
            "score": round_score(process_score),
            "weight": methodology.weights.process,
            "weighted_score": round_score(process_score * methodology.weights.process),
            "inputs": {
                field.id: round_score(clamp(process_values.get(field.id, 0.0))) for field in process_model.fields
            },
        },
        "integrity": {
            "label": methodology.integrity.label,
            "score": round_score(integrity_score),
            "weight": methodology.weights.integrity,
            "weighted_score": round_score(integrity_score * methodology.weights.integrity),
            "components": {
                key: round_score(clamp(integrity_values.get(key, 0.0))) for key in methodology.integrity.components
            },
        },
        "penalties": {
            "total": round_score(total_penalties),
            "items": [p.__dict__ for p in penalties],
        },
    }

    return ScoreResult(
        product=methodology.product,
        methodology_id=methodology.id,
        methodology_version=methodology.version,
        quality_score=round_score(quality_score),
        process_score=round_score(process_score),
        integrity_score=round_score(integrity_score),
        penalties=round_score(total_penalties),
        total_score=round_score(total_score),
        confidence_level=confidence_level,
        classification=classification,
        interpretation=interpretation,
        components=components,
    )
